---
title: "Testing bi-allelic system"
output:
  html_document: default
---


##Introduction
the goal of this notebook is to check wether the selection coefficient which is a selection model for bi-allelic system, is working correctly. To do so, we perform two set of test. First, we simulate a population with a small advantage for a gene a over A, but where A is more present than a at the begining. We compute the probability that gene a or A is fixed. We compare this to coefficient of the transition matrix for infinite time.   
In the second part, we follow the same analysis, but instead of changing the initial frequency, we change the advantage given to each allele and its dominance. We distinguish cases where one of the allele fix to cases where a polymorph system is favored. 

##Changing frequency
### General input of the folder
```{r}
devtools::load_all(pkg="../RQuantiNemo")
my_sim <- new("simulation", sim.dir = "selection_coefficient/")
my_sim  <- setParameter(my_sim,"replicates",50)
my_sim  <- setParameter(my_sim,"generations",250)
my_sim  <- setParameter(my_sim,"patch_capacity",100)
my_sim  <- setParameter(my_sim,"quanti_loci",1)
my_sim  <- setParameter(my_sim,"quanti_all",2)
my_sim  <- setParameter(my_sim,"quanti_dominance_model",1)
my_sim  <- setParameter(my_sim,"patch_coef_sel",0.1)
my_sim  <- setParameter(my_sim,"quanti_selection_model",4)
my_sim  <- setParameter(my_sim,"stat","{q.adlt.nbAll q.adlt.ho q.meanG_p }")
```


###create the genotype
```{r}
getGenotype <- function(freq,pop){
  n1 = 2*round((1-freq)*pop)
  n2 = pop-n1
  print(n2)
  return(data.frame(col_1 = rep(1,pop),col_2 = c(rep(12,n1),rep(22,n2))))
}
```


###loop overfrequency
```{r}
freqs = c(1,0.99,0.98,0.97,0.96,0.95,0.9,0.85,0.8,0.75,0.7)
#freqs = c(1,0.9,0.8,0.7,0.6,0.5)
for (freq in freqs)
{
  my_sim@sim.name <- as.character(freq)
  alleles <- data.frame(col_locus = c(1,1),col_allele = c(1,2),col_allelic_value = c(0, 1), col_ini_freq = c(1-freq, freq))
  my_sim <-  addFile(my_sim, "quanti_allelic_file",alleles)
  genotype <- getGenotype(freq, my_sim@parameters[["patch_capacity"]])
  my_sim <-  addFile(my_sim, "quanti_ini_genotypes",genotype )
  result <- run(my_sim, verbose = F)
}
```


###Function to compute the transition matrix
```{r}
computeTransition <- function(pop.size = 100, s = 0.1)
{
  trans <- matrix(rep(0,(pop.size+1)^2),nrow = (pop.size+1))
  for(i in 0:pop.size)
  {
    p <-  i/pop.size
    q <-  1-p
    omega = p^2 +2*p*q + (1-s)*q^2
    pp <-  p/omega # = p^2/omega + 1/2*2*p*q/omega
    trans[i+1,] <- dbinom(0:pop.size,pop.size,pp)
  }
  for (i in(1:10)) {trans <- trans%*%trans}
  return(trans)
}
trans <- computeTransition()
```


###Loading and ploting the date from output file
The result are very similar, though we can see that the points are in general just above the curve. 
It's probably coming for a bias from the estimator. 
```{r}
freqs = c(1,0.99,0.98,0.97,0.96,0.95,0.9,0.85,0.8,0.75,0.7)
gens=list()
plot(0:39,trans[1:40, 1],type="l")
for (freq in freqs){
  gen <- read.table(paste("selection_coefficient/",as.character(freq), '/simulation_mean.txt',sep=""),header=T)
  gens[[as.character(freq)]] <- gen$q.meanG[gen$generation == 250]/2
  points((1-freq)*100,gen$q.meanG[gen$generation == 250]/2)
}
stopifnot(sum((unlist(gens) - trans[1+(1-freqs)*100])^2)<0.1)
```



##Changing selection strength
### General input of the folder
```{r}
devtools::load_all(pkg="../RQuantiNemo")
my_sim <- new("simulation", sim.dir = "selection_coefficient/")
my_sim  <- setParameter(my_sim,"replicates",1)
my_sim  <- setParameter(my_sim,"generations",100)
my_sim  <- setParameter(my_sim,"patch_capacity",10000)
my_sim  <- setParameter(my_sim,"quanti_loci",1)
my_sim  <- setParameter(my_sim,"quanti_all",2)
my_sim  <- setParameter(my_sim,"quanti_dominance_model",1)
my_sim  <- setParameter(my_sim,"quanti_selection_model",4)
my_sim  <- setParameter(my_sim,"seed","{1654,7894}")
my_sim  <- setParameter(my_sim,"quanti_save_genotype",0)
my_sim  <- setParameter(my_sim,"stat","{q.adlt.nbAll q.adlt.ho q.meanG_p meanW_p}")
```


###loop overfrequency
```{r}
freq = 0.99
ss = seq(0,1,0.25)
hs = seq(0,1,0.25)
for (s in ss){
  for (h in hs){
  my_sim@sim.name <- paste("s_",as.character(s),"h_",as.character(h),sep = "")
  alleles <- data.frame(col_locus = c(1,1),col_allele = c(1,2),col_allelic_value = c(0, 1), col_ini_freq = c(1-freq, freq))
  my_sim <-  addFile(my_sim, "quanti_allelic_file",alleles)
  my_sim  <- setParameter(my_sim,"patch_coef_sel",s)
  my_sim  <- setParameter(my_sim,"quanti_dominance_mean",h)
  result <- run(my_sim, verbose = F)
  }
}
```

###Loading and ploting the date from output file
The result are very similar, though we can see that the points are in general just above the curve. 
It's probably coming for a bias from the estimator. 
```{r}
fitness <- function(p,s,h){1-s+2*s*(1-h)*p-s*(1-2*h)*p^2}
p.new <- function(p,s,h) {p+p*(1-p)*s/fitness(p,s,h)*(1-h-(1-2*h)*p)} # The mathematical theory of selection, recombination and mutation: R. Burger: page 32
res <- c()
for (s in ss){
  for (h in hs){
    my_sim@sim.name <- paste("s_",as.character(s),"h_",as.character(h),sep = "")
    stat <- loadStat(my_sim)
    res = c(res,tail(stat$meanW,1))
    evol = c(0.01)
    for(i in 1:100){evol = c(evol,p.new(tail(evol,1),s,h))}
    plot(fitness(evol,s,h),main =  paste("s = ",as.character(s),", h = ",as.character(h),sep = ""),type = "l")
    points(stat$meanW)
  }
}
```
###Checking if distance is ok for some parameters
```{r}

distance <- function(s,h)
{
    my_sim@sim.name <- paste("s_",as.character(s),"h_",as.character(h),sep = "")
    stat <- loadStat(my_sim)
    res = c(res,tail(stat$meanW,1))
    evol = c(0.01)
    for(i in 1:99){evol = c(evol,p.new(tail(evol,1),s,h))}
    return (sum(fitness(evol,s,h)-stat$meanW)^2)
}
stopifnot(distance(0,0)==0)
stopifnot(distance(0,1)==0)
stopifnot(distance(1,0)<0.01)
stopifnot(distance(1,1)<1)
stopifnot(distance(0.25,0.25)<1)
stopifnot(distance(0.25,0.75)<0.1)
stopifnot(distance(0.5,1)<0.1)
stopifnot(distance(0.25,0)<0.1)
stopifnot(distance(0,0.25)<0.1)
stopifnot(distance(1,0.75)<0.1)
```

##Favoring heterozygocity
Since we already did intensiv test above about the set of parameters and if the selection is acting at the rifgt space, we do a much simpler test here, just checking that the population remain with a polymorphism eventhough drift should fix a single allele if h < 0. 
```{r}
devtools::load_all(pkg="../RQuantiNemo")
my_sim <- new("simulation", sim.dir = "selection_coefficient/")
my_sim  <- setParameter(my_sim,"replicates",5)
my_sim  <- setParameter(my_sim,"generations",500)
my_sim  <- setParameter(my_sim,"patch_capacity",100)
my_sim  <- setParameter(my_sim,"quanti_loci",1)
my_sim  <- setParameter(my_sim,"quanti_all",2)
my_sim  <- setParameter(my_sim,"quanti_dominance_model",1)
my_sim  <- setParameter(my_sim,"quanti_selection_model",4)
my_sim  <- setParameter(my_sim,"seed","{1654,7894}")
my_sim  <- setParameter(my_sim,"quanti_save_genotype",2)
my_sim  <- setParameter(my_sim,"stat","{q.adlt.nbAll q.adlt.ho q.meanG_p meanW_p}")
```

```{r}
freq = 0.5
h = -0.50
s = 0.5
my_sim@sim.name <- paste("s_",as.character(s),"h_",as.character(h),sep = "")
alleles <- data.frame(col_locus = c(1,1),col_allele = c(1,2),col_allelic_value = c(0, 1), col_ini_freq = c(1-freq, freq))
my_sim <-  addFile(my_sim, "quanti_allelic_file",alleles)
my_sim  <- setParameter(my_sim,"patch_coef_sel",s)
my_sim  <- setParameter(my_sim,"quanti_dominance_mean",h)
result <- run(my_sim, verbose = F)
my_sim@sim.name <- paste("s_",as.character(s),"h_",as.character(h),sep = "")
stat <- loadStat(my_sim)
res = c(res,tail(stat$q.adlt.nbAll,1))
plot(stat$q.adlt.nbAll)
stopifnot(tail(stat$q.adlt.nbAll,1)==2)
```
